### ISSUE TITLE
i3status-2.12 memory leak (2019)

### ISSUE THREAD

**Issue by gamma62** [i3/i3status#330]
I saw in top that i3status uses 3GiB RAM, after reload the value dropped down to some MiB.
But then it increased again:
$ while true; do top -b | awk '/i3status/ {print $6,$7,$NF}' ; sleep 10; done
110M 18M i3status
111M 18M i3status
111M 18M i3status
112M 18M i3status
112M 18M i3status
113M 18M i3status
113M 18M i3status
114M 19M i3status
114M 19M i3status
114M 19M i3status

I see that this is not the first mention of this problem. 
How can I find the root cause? Coverity? Valgrind?  Or is there a patch somewhere? 
I'm using FreeBSD 12.0

regards,
A

**Comment by gamma62** [comment]
I read some threads. Here is the minimal config. This seems to be OK, no leak.

general {
        output_format = i3bar
        colors = true
        interval = 5
}
order += "tztime local"
tztime local {
        format = " %H:%M %a "
#        locale = "hu_HU.UTF-8"
}

But when I added locale (not the default!) to the tztime config, then 
the memory usage began to increase.
Other modules (battery, ethernet, wireless) seem to be OK, at least in my quick tests.

**Comment by Airblader** [comment]
That's interesting information already, thank you. It's a bit surprising, though, since we really only call `setlocale` and that call seems fine to me.

Could you maybe use valgrind to figure out where memory gets lost? I haven't used it in a while, but something like `valgrind --leak-check=yes i3status` should work.

**Comment by gamma62** [comment]
Terrible formatting... replacing LC_ALL with LC_TIME is the key.

Anyway, calling setlocale(LC_ALL, locale); setlocale(LC_ALL, ""); in a loop should not leak. So this is not really related to i3status. The other option would be to handle locale setting like TZ. But that may have side effects in other modules.

I'm satisfied with this patch.

**Comment by Airblader** [comment]
Yeah, this doesn't really seem like an i3status issue. I wonder if other FreeBSD users can reproduce this?
